src/lib.rs

```diff
@@ -6,18 +6,203 @@ An experimental x86_64 bootloader that works on both BIOS and UEFI systems.
+struct DiskImageFile {
+    source: PathBuf,
+    destination: String,
+}
+
+/// DiskImageBuilder helps create disk images for a specified set of files.
+/// It can currently create MBR (BIOS), GPT (UEFI), and TFTP (UEFI) images.
+pub struct DiskImageBuilder {
+    files: Vec<DiskImageFile>,
+}
+
+impl DiskImageBuilder {
+    /// Create a new instance of DiskImageBuilder, with the specified kernel.
+    pub fn new(kernel: &Path) -> Self {
+        let mut obj = Self::empty();
+        obj.set_kernel(kernel);
+        obj
+    }
+
+    /// Create a new, empty instance of DiskImageBuilder
+    pub fn empty() -> Self {
+        Self { files: Vec::new() }
+    }
+
+    /// Add or replace a kernel to be included in the final image.
+    pub fn set_kernel(&mut self, path: &Path) -> &mut Self {
+        self.add_or_replace_file(path, KERNEL_FILE_NAME)
+    }
+
+    /// Add or replace a ramdisk to be included in the final image.
+    pub fn set_ramdisk(&mut self, path: &Path) -> &mut Self {
+        self.add_or_replace_file(path, RAMDISK_FILE_NAME)
+    }
+
+    /// Configures the runtime behavior of the bootloader.
+    pub fn set_boot_config(&mut self, boot_config: &BootConfig) -> &mut Self {
+        let json =
+            serde_json::to_string_pretty(boot_config).expect("failed to serialize BootConfig");
+        let bytes = json.as_bytes();
+        self.add_or_replace_file_byte_content(bytes, CONFIG_FILE_NAME)
+    }
+
+    /// Add or replace a file from a byte array
+    pub fn add_or_replace_file_byte_content(&mut self, data: &[u8], target: &str) -> &mut Self {
+        let temp_path = temp_dir();
+        let file_name = temp_path.join("bytes.tmp");
+        fs::create_dir_all(temp_path).expect("Failed to create temp directory");
+        let mut temp_file =
+            fs::File::create(file_name.clone()).expect("Failed to create temp file");
+        temp_file
+            .write_all(data)
+            .expect("Failed to write data to temp file");
+        temp_file
+            .sync_all()
+            .expect("Failed to flush temp file to disk");
+
+        self.add_or_replace_file(&file_name, target)
+    }
+
+    /// Add or replace arbitrary files.
+    /// NOTE: You can overwrite internal files if you choose, such as EFI/BOOT/BOOTX64.EFI
+    /// This can be useful in situations where you want to generate an image, but not use the provided bootloader.
+    pub fn add_or_replace_file(&mut self, path: &Path, target: &str) -> &mut Self {
+        self.files.insert(
+            0,
+            DiskImageFile {
+                source: path.to_path_buf(),
+                destination: target.to_string(),
+            },
+        );
+        self
+    }
+    fn create_fat_filesystem_image(
+        &self,
+        internal_files: BTreeMap<&str, &Path>,
+    ) -> anyhow::Result<NamedTempFile> {
+        let mut local_map = BTreeMap::new();
+
+        for k in internal_files {
+            local_map.insert(k.0, k.1);
+        }
```

Comment:
We should set the internal files after `self.files`, otherwise the user supplied files might overwrite some core files.

Ideally we should also report some error in case there conflicts. Alternatively, we could place all custom files in a subfolder.

Reply:
This was an intentional feature. I will undo it as discussed in the reply below.

As for the error, I'm not sure here personally. I'll go whatever direction you like.
My thought here was: 
During build, we have some default files, we loop and add them.
Later in the build, we have "override" files that we replace these with.

So, with that thought process:
1. Allow overwriting of own files
2. Error when overwriting bootloader files?